[sw/examples] Turn hello_world into a functest

pamaury opened the pull request:
There are currently no clear instruction on how to use this example.
Those that exists are scattered in the documentation (FPGA) and for
verilator the only good reference is the english breakfast verilator
script. By adding a rule, it makes it much easier to actually
run it without manual fiddling. Unfortunately, a test is still
not quite right because hello_world never returns so this will make
the "test" fails. This is why we use the previously introduced
notion of example. This example can be run by using
```./bazelisk.sh run //sw/device/examples/hello_world:hello_world_example_fpga_cw310_test_rom```
and
```./bazelisk.sh run //sw/device/examples/hello_world:hello_world_example_sim_verilator```

a-will commented:
This isn't a test, though, and I don't think that's the appropriate model in the build system :thinking:.  IMHO, bazel doesn't need to have a runner for every target in the code base, and our tools need to be useful outside of bazel instrumentation.

May I recommend creating clearer documentation of how to use opentitantool to "run" or "load" an example binary, depending on the target platform, in addition to how to locate the needed tools and artifacts?

If you and others want bazel instrumentation anyway, then creating a separate non-test rule is probably the way to go. I suspect much of `opentitan_functest` could be refactored to be common, but likely we'd want to use `native.sh_binary` instead of `native.sh_test` (assuming a shim is still desired to handle setup).

pamaury commented:
Yes I am aware of why `functest` is not a good model for this. My plan is to modify it to make it able to create a `sh_binary` with no timeout so one can just `run` it. This was a test PR to see what this breaks because of the renaming of the targets (the opentitan_functest introduce an extra `prog` in the name). I still think being able to run it with bazel is valuable compared to running it with manual operations.

pamaury commented:
I have been digging a little bit more and one issue I have found is that even if I generate a `sh_binary`, it has to depend on the bitstream. However, `//hw/bitstream:rom_with_fake_keys` is marked as test only which makes it impossible to use it. @a-will any ideas? Changing test only on the bitstream could be a big change, I am not sure what impact it would have and why it was marked as test only.

a-will commented:
> I have been digging a little bit more and one issue I have found is that even if I generate a `sh_binary`, it has to depend on the bitstream. However, `//hw/bitstream:rom_with_fake_keys` is marked as test only which makes it impossible to use it. @a-will any ideas? Changing test only on the bitstream could be a big change, I am not sure what impact it would have and why it was marked as test only.

@alphan might be best able to give the explanation, as the infection from the testonly virus ostensibly happened at https://github.com/lowRISC/opentitan/pull/16221 . It looks to me like a well-meaning attempt to prevent fake keys from making their way into production releases, but the reality is that this repo's releases are almost entirely about test platforms. Even the targets in `//release` are marked with `testonly = True`.

So, you could choose to do the same here. If you add `testonly = True` to the `sh_binary` target, then it can use the bitstreams. Despite the name of the attribute, it doesn't mean you can't "bazel run" the target. It merely adds an additional requirement for edges in the build graph.

Alternatively, we could reconsider whether the testonly attribute is working as intended and if it's worth the upkeep.

pamaury committed:
[baze] Add an opentitan_example rule/macro

The intent of this rule is to create a runnable executable that is
not a test and so has no pass/fail or timeout.

Signed-off-by: Amaury Pouly <amaury.pouly@lowrisc.org>

pamaury committed:
[examples] Make hello_world an opentitan_example

Signed-off-by: Amaury Pouly <amaury.pouly@lowrisc.org>

rswarbrick commented:
Hi @pamaury! Is this PR still current? If so, should it be rebased? If not, maybe it should be closed?

pamaury commented:
I am closing, there hasn't been much interest in this for a long time
